bitcoin/src/blockdata/transaction.rs

```diff
@@ -993,6 +993,83 @@ impl Transaction {
         // `Transaction` docs for full explanation).
         self.input.is_empty()
     }
+
+    /// Returns a reference to the input at `input_index` if it exists.
+    #[inline]
+    pub fn tx_in(&self, input_index: usize) -> Result<&TxIn, InputsIndexError> {
+        self.input
+            .get(input_index)
+            .ok_or(IndexOutOfBoundsError { index: input_index, length: self.input.len() }.into())
+    }
+
+    /// Returns a reference to the output at `output_index` if it exists.
+    #[inline]
+    pub fn tx_out(&self, output_index: usize) -> Result<&TxOut, OutputsIndexError> {
+        self.output
+            .get(output_index)
+            .ok_or(IndexOutOfBoundsError { index: output_index, length: self.output.len() }.into())
+    }
+}
+
+/// Error attempting to do an out of bounds access a the transaction inputs vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct InputsIndexError(pub IndexOutOfBoundsError);
+
+impl fmt::Display for InputsIndexError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write_err!(f, "invalid input index"; self.0)
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for InputsIndexError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
+}
+
+impl From<IndexOutOfBoundsError> for InputsIndexError {
+    fn from(e: IndexOutOfBoundsError) -> Self { Self(e) }
+}
+
+/// Error attempting to do an out of bounds access a the transaction outputs vector.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct OutputsIndexError(pub IndexOutOfBoundsError);
+
+impl fmt::Display for OutputsIndexError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write_err!(f, "invalid output index"; self.0)
```

Comment:
Either this or the `source` implementation is wrong - in `std` it will never display the inner information.

You should use `write` and directly access the fields of the inner error to improve the error message because just switching to `write!` would create this long redundant message:

> invalid output index: index {} is out-of-bounds for vector with length {}

But anyway, I think that this approach is either confusing or creates needlessly long message if you return `source`, so I'm not really entirely happy about it (but it may be the least bad option, didn't see the rest of the code yet).

Reply:
Excluding that I was very much wrong returning `None` from `source`, the error message is as I expected it to be. What is confusing about it?
E.g with numbers:
`invalid output index: index 2 is out-of-bounds for vector with length 1`

Reply:
There are two options to impl `Error`: return `Some` and impl display, which produces long messages, or delegate everything to the inner error. The latter is confusing in that if you do non-trivial inspecting of `source()` somewhere up the chain this will skip the wrapped type, so you can't downcast to it. I guess long error messages might be the lesser evil.